tests/test_updater_top_level_update.py

```diff
@@ -0,0 +1,462 @@
+        # remote root (v2)
+        self.sim.root.version += 1
+        self.sim.publish_root()
+
+        self._run_refresh()
+
+        self._assert_files_exist(TOP_LEVEL_ROLE_NAMES)
+        for role in TOP_LEVEL_ROLE_NAMES:
+            version = 2 if role == "root" else None
+            self._assert_content_equals(role, version)
+
+    def test_trusted_root_missing(self) -> None:
+        os.remove(os.path.join(self.metadata_dir, "root.json"))
+        with self.assertRaises(OSError):
+            self._run_refresh()
+
+        # Metadata dir is empty
+        self.assertFalse(os.listdir(self.metadata_dir))
+
+    def test_trusted_root_expired(self) -> None:
+        # Create an expired root version
+        self.sim.root.expires = self.past_datetime
+        self.sim.root.version += 1
+        self.sim.publish_root()
+
+        # Update to latest root which is expired but still
+        # saved as a local root.
+        updater = self._init_updater()
+        with self.assertRaises(ExpiredMetadataError):
+            updater.refresh()
+
+        self._assert_files_exist(["root"])
+        self._assert_content_equals("root", 2)
+
+        # The expiration of the trusted root metadata file does not lead
+        # to failure and Updater in successfully initialized.
+        updater = self._init_updater()
+
+        # Create a non-expired root version and refresh
+        self.sim.root.expires = self.sim.safe_expiry
+        self.sim.root.version += 1
+        self.sim.publish_root()
+        updater.refresh()
+
+        # Root is successfully updated to latest version
+        self._assert_files_exist(TOP_LEVEL_ROLE_NAMES)
+        self._assert_content_equals("root", 3)
+
+    def test_trusted_root_unsigned(self) -> None:
+        # Local trusted root is not signed
+        root_path = os.path.join(self.metadata_dir, "root.json")
+        md_root = Metadata.from_file(root_path)
+        md_root.signatures.clear()
+        md_root.to_file(root_path)
+
+        with self.assertRaises(UnsignedMetadataError):
+            self._run_refresh()
+
+        # The update failed, no changes in metadata
+        self._assert_files_exist(["root"])
+        md_root_after = Metadata.from_file(root_path)
+        self.assertEqual(md_root.to_bytes(), md_root_after.to_bytes())
+
+    def test_max_root_rotations(self) -> None:
+        # Root must stop looking for new versions after Y number of
+        # intermediate files were downloaded.
+        updater = self._init_updater()
+        updater.config.max_root_rotations = 3
+
+        # Create some number of roots greater than 'max_root_rotations'
+        while self.sim.root.version < updater.config.max_root_rotations + 3:
+            self.sim.root.version += 1
+            self.sim.publish_root()
+
+        root_path = os.path.join(self.metadata_dir, "root.json")
+        md_root = Metadata.from_file(root_path)
+        initial_root_version = md_root.signed.version
+
+        updater.refresh()
```

Comment:
Surprisingly (for me too), this is the correct behaviour. The spec treats this case the same way as if the next root version was not found. The legacy code does the same thing.

>  If this file is not available, or we have downloaded more than Y number of root metadata files (because the exact number is as yet unknown), then go to step 5.3.10.

Reply:
that seems like a very weird choice... thanks for checking